lib/SILGen/SILGenPoly.cpp

```diff
@@ -1619,6 +1621,10 @@ static ManagedValue applyTrivialConversions(SILGenFunction &SGF,
                                             SILType outerType) {
   auto innerASTTy = innerValue.getType().getASTType();
   auto outerASTTy = outerType.getASTType();
+  if (innerASTTy->hasArchetype())
```

Comment:
Note: this change is necessary to avoid a crash in `SILGenModule::getOrCreateCustomDerivativeThunk` when calling `forwardFunctionArguments`:

```swift
import _Differentiation

func foo<T: Differentiable>(_ x: T) -> T { x }

@derivative(of: foo)
func vjpFoo<T: Differentiable>(_ x: T) -> (value: T, pullback: (T.TangentVector) -> T.TangentVector) {
  fatalError()
}
```
```console
$ swiftc -Xfrontend -enable-experimental-differentiable-programming -emit-silgen okay.swift
innerASTTy:
(primary_archetype_type address=0x7f87b4116f90 conforms_to=_Differentiation.(file).Differentiable name=τ_0_0

  (nested_type=TangentVector <<unresolved>>))

outerASTTy:
(generic_type_param_type depth=0 index=0)

unhandled reabstraction type mismatch
UNREACHABLE executed at /Users/danielzheng/swift-master/swift/lib/SILGen/SILGenPoly.cpp:1663!
Stack dump:
0.	Program arguments: /Users/danielzheng/swift-master/build/Ninja-ReleaseAssert/swift-macosx-x86_64/bin/swift -frontend -emit-silgen -primary-file okay.swift -target x86_64-apple-darwin18.7.0 -enable-objc-interop -sdk /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk -color-diagnostics -enable-experimental-differentiable-programming -module-name okay -o -
1.	Swift version 5.3-dev (LLVM 97ca3af693, Swift 59a9fb3b55)
2.	While evaluating request SILGenSourceFileRequest(SIL Generation for file "okay.swift")
3.	While emitting SIL for 'vjpFoo(_:)' (at okay.swift:6:1)
4.	While silgen emitFunction SIL function "@$s4okay6vjpFooyx5value_13TangentVectorQzAEc8pullbacktx16_Differentiation14DifferentiableRzlF".
 for 'vjpFoo(_:)' (at okay.swift:6:1)
0  swift                    0x00000001060818e5 llvm::sys::PrintStackTrace(llvm::raw_ostream&) + 37
1  swift                    0x0000000106080898 llvm::sys::RunSignalHandlers() + 248
2  swift                    0x0000000106081edc SignalHandler(int) + 268
3  libsystem_platform.dylib 0x00007fff6deebb5d _sigtramp + 29
4  swift                    0x00000001075243c2 cmark_strbuf_put.cold.1 + 317266
5  libsystem_c.dylib        0x00007fff6dda56a6 abort + 127
6  swift                    0x000000010749a9ae llvm::llvm_unreachable_internal(char const*, char const*, unsigned int) + 462
7  swift                    0x0000000102041d35 applyTrivialConversions(swift::Lowering::SILGenFunction&, swift::SILLocation, swift::Lowering::ManagedValue, swift::SILType) + 821
8  swift                    0x0000000102038836 forwardFunctionArguments(swift::Lowering::SILGenFunction&, swift::SILLocation, swift::CanTypeWrapper<swift::SILFunctionType>, llvm::ArrayRef<swift::Lowering::ManagedValue>, llvm::SmallVectorImpl<swift::SILValue>&) + 390
9  swift                    0x0000000102037a00 swift::Lowering::SILGenModule::getOrCreateCustomDerivativeThunk(swift::SILFunction*, swift::SILFunction*, swift::AutoDiffConfig const&, swift::AutoDiffDerivativeFunctionKind) + 2848
10 swift                    0x0000000101f779de swift::Lowering::SILGenModule::emitDifferentiabilityWitness(swift::AbstractFunctionDecl*, swift::SILFunction*, swift::AutoDiffConfig const&, swift::SILFunction*, swift::SILFunction*, swift::DeclAttribute const*) + 526
11 swift                    0x0000000101f7776f swift::Lowering::SILGenModule::emitDifferentiabilityWitnessesForFunction(swift::SILDeclRef, swift::SILFunction*)::$_1::operator()(swift::DeclAttributes&) const + 1103
12 swift                    0x0000000101f7730c swift::Lowering::SILGenModule::emitDifferentiabilityWitnessesForFunction(swift::SILDeclRef, swift::SILFunction*) + 172
13 swift                    0x0000000101f77245 swift::Lowering::SILGenModule::postEmitFunction(swift::SILDeclRef, swift::SILFunction*) + 245
14 swift                    0x0000000101f7e4c8 swift::Lowering::SILGenModule::emitFunction(swift::FuncDecl*)::$_3::operator()(swift::SILFunction*) const + 264
```

```
Failing Tests (2):
    Swift(macosx-x86_64) :: AutoDiff/SILGen/sil_differentiability_witness_silgen.swift
    Swift(macosx-x86_64) :: AutoDiff/Serialization/derivative_attr.swift
```

---

@jckarter: I wonder if this change is reasonable? It feels bad.

I looked a bit into changing other code (`SILFunctionType::getAutoDiffDerivativeFunctionType`) so that this change isn't necessary but couldn't figure it out. Maybe others have ideas.